Changed the signal handling to named pipes in the zoo test

raylu-hdf opened the pull request:
For communication between the writer and reader, mainly in the zoo test as a starting point.  The purpose is to make easy porting to Windows.

Commit:
Changed the signal handling to named pipes for communication between the writer and reader, mainly in the zoo test as a starting point.

raylu-hdf committed:
Merge branch 'feature/vfd_swmr' into raylu_zoo_test

raylu-hdf commented on test/genall5.c:
```diff
@@ -2755,12 +2755,13 @@ tend_zoo(hid_t fid, const char *base_path, struct timespec *lastmsgtime,
     }
 out:
     if (!ok) {
```
The original code set the `lastmsgtime` to 0.  `below_speed_limit` would always return true.  I changed it to use the `lastmsgtime` passed into this function, which is set by the caller.

gnuoyd replied:
I think that you have misunderstood `below_speed_limit`. It's a rate-limiting utility. Let me know how I can clarify the comment on `below_speed_limit`.

raylu-hdf replied:
This function is no longer used by the zoo test after refactoring the code.  So I will close this comment.

raylu-hdf commented on test/testvfdswmr.sh.in:
```diff
@@ -571,24 +571,21 @@ done
 # read and written by VFD SWMR.
 #
 if [ ${do_zoo:-no} = yes ]; then
-	[ -e ./fifo ] && rm -f ./fifo
```
The C program does these two steps now: removing and creating the named pipes.

raylu-hdf commented on test/testvfdswmr.sh.in:
```diff
@@ -571,24 +571,21 @@ done
-	STDIN_PATH="./fifo" catch_out_err_and_rc vfd_swmr_zoo_writer \
-	    ../vfd_swmr_zoo_writer -m 1000 -q &
+	catch_out_err_and_rc vfd_swmr_zoo_writer \
+	    ../vfd_swmr_zoo_writer -l 5 -m 1000 -q &
```
`-l 5` tells the writer that it has 5 seconds to finish the creation or deletion of objects. `-l 10` tells the reader that it has 10 seconds to validate the creation or deletion of objects.

gnuoyd replied:
Careful setting tight time limits, slow or overburdened machines will run right over the limits and produce time-wasting spurious failures. A time limit one minute or longer is probably more appropriate.

raylu-hdf replied:
`-l` is the maximal interval (between the writer's finishing zoo creation or deletion and reader's finishing validation) in terms of tick now.

raylu-hdf commented on test/testvfdswmr.sh.in:
```diff
@@ -571,24 +571,21 @@ done
 	# reader will find the shadow file when it opens
 	# the .h5 file.
 	wait $pid_reader
-	kill -USR1 $(cat vfd_swmr_zoo_writer.pid)
```
I don't know why this is necessary.  The test works fine without it.

gnuoyd replied:
You deleted the part of the writer that waits for the signal.

raylu-hdf commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
-        /* Avoid deadlock: flush the file before waiting for the reader's
-         * message.
-         */
-        if (H5Fflush(fid, H5F_SCOPE_GLOBAL) < 0)
```
The test hangs with this function call.  I'm not sure why.

vchoi-hdfgroup commented on test/vfd_swmr_common.c:
```diff
@@ -68,6 +68,21 @@ below_speed_limit(struct timespec *last, const struct timespec *ival)
+ *
+ * This routine may quietly perform a too-short sleep if an error occurs
+ * in nanosleep(2).
+ */
```
Don't quite understand why this new decisleep() is not using H5_nanosleep() in the library?

raylu-hdf replied:
I fixed it.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -108,6 +111,9 @@ vfd_swmr_writer_may_increase_tick_to(uint64_t new_tick, bool wait_for_reader)
     dbgf(3, "%s: enter\n", __func__);
 
     if (fd == -1) {
+        if (access("./shared_tick_num", F_OK ) < 0)
```
Use HDaccess().

raylu-hdf replied:
It's no longer used.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
+        if (notify != 3)
+            errx(EXIT_FAILURE, "expected 2 but read %d", notify);
+
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) == -1)
```
Use HDclock_gettime()

raylu-hdf replied:
Fixed.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
 
         ostate = initstate(seed, vector, _arraycount(vector));
 
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
```
Use HDclock_gettime().

raylu-hdf replied:
Fixed.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -259,6 +270,18 @@ main(int argc, char **argv)
+        case 'l':
+            /* Expected maximal time for reader's validation of zoo creation or deletion */
+            errno = 0;
+            tmpl = strtoul(optarg, &end, 0);
```
Use HDstrtoul().

raylu-hdf replied:
Changed.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -33,6 +33,8 @@
 #define _arraycount(_a) (sizeof(_a)/sizeof(_a[0]))
```
Similar definition of __arraycount exists in vfd_swmr.c.  Is it possible to put it in a common header or file so both can use access it?

raylu-hdf replied:
It's not used in the zoo test.  I'll take it out.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
-         */
-        if (H5Fflush(fid, H5F_SCOPE_GLOBAL) < 0)
-            errx(EXIT_FAILURE, "%s: H5Fflush failed", __func__);
+        /* Notify the reader of finishing zoo creation */
```
Write description/comments about how pipes are used in coordinating between the writer and reader

raylu-hdf replied:
I'll improve the comments.

gnuoyd commented on test/genall5.c:
```diff
@@ -2755,12 +2755,13 @@ tend_zoo(hid_t fid, const char *base_path, struct timespec *lastmsgtime,
+            if (below_speed_limit(lastmsgtime, &config.msgival)) {
+                last_failure_mssg = failure_mssg;
+                warnx("%s: %s", __func__, failure_mssg);
+            }
```
I don't understand the importance of this change.

The VFD SWMR reader will produce a deluge of identical error messages while it waits for the content it expects to appear in the HDF5 file that it shares with the writer, unless the error messages are rate-limited. The non-VFD SWMR "zoo" tests are not expected to ever trigger rate-limiting, so limiting the output rate, anyway, does not affect the error output in any significant way.

raylu-hdf replied:
We talked about it in our meeting.  It should be ok now.

gnuoyd commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -108,6 +111,9 @@ vfd_swmr_writer_may_increase_tick_to(uint64_t new_tick, bool wait_for_reader)
     if (fd == -1) {
+        if (access("./shared_tick_num", F_OK ) < 0)
+            return true;
+
```
I think that if you need an `access(2)` call, it deserves an explanatory comment. An access check is inherent in the next statement, the `open(2)` call.

raylu-hdf replied:
When the file doesn't exist yet, `open` will spill out an error message.

gnuoyd commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -324,14 +347,26 @@ main(int argc, char **argv)
+            errx(EXIT_FAILURE, "HDmkfifo");
+
+        if (HDmkfifo(fifo_reader_to_writer, 0666) < 0)
+            errx(EXIT_FAILURE, "HDmkfifo");
```
Maybe use 0600 here per the "principle of least authority" ?

raylu-hdf replied:
You're right.  I fixed it.

gnuoyd commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
 
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
+            errx(EXIT_FAILURE, "%s: clock_gettime", __func__);
+
```
What are these`lastmsgtime` re-initializations for?

raylu-hdf replied:
We talked about it in our meeting.

gnuoyd commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
-            err(EXIT_FAILURE, "write");
+        /* Make sure zoo validation doesn't take longer than the expected time */
+        if (below_speed_limit(&last, &ival))
+            warnx("validate_zoo took too long to finish");
```
This isn't really how `below_speed_limit` is meant to be used. :-) You need a function like [`timespeccmp`](https://man.netbsd.org/timespeccmp).

raylu-hdf replied:
I think a new function will be similar to `below_speed_limit `

raylu-hdf committed:
Added the HDF5 prefix (HD_) to several system calls.

raylu-hdf committed:
Merge branch 'feature/vfd_swmr' of https://github.com/HDFGroup/hdf5 into raylu_zoo_test

raylu-hdf committed:
Merge branch 'raylu_zoo_test' of https://github.com/HDFGroup/hdf5 into raylu_zoo_test

raylu-hdf committed:
Added the prefix (HD) to several C functions.

raylu-hdf committed:
Two main changes include: re-arranging the communication between the writer and reader through the named pipes; using the error report consistent with other tests.

raylu-hdf committed:
Removed the two functions (vfd_swmr_writer_may_increase_tick_to and vfd_swmr_reader_did_increase_tick_to) that overrode the functions of the same names in the SWMR driver.
Also added a command-line option to disable named pipes so that the writer and reader can run seperately.

raylu-hdf committed:
Removed an unused macro definition.

raylu-hdf commented:
This PR is ready for another round of review, hopefully the final one.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -74,335 +57,437 @@ print_cache_hits(H5C_t H5_ATTR_UNUSED *cache)
     exit(EXIT_FAILURE);
 }
 
-bool
```
Should we also remove the corresponding library instrumentation function of this routine?

raylu-hdf replied:
Dana will handle it.

vchoi-hdfgroup commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -74,335 +57,437 @@ print_cache_hits(H5C_t H5_ATTR_UNUSED *cache)
-    return false;
-}
-
-void
```
Should we also remove the corresponding library instrumentation function of this routine?

raylu-hdf replied:
That's a good question.  I'll ask John about it in tomorrow's meeting.

raylu-hdf replied:
Dana will get rid of them.

raylu-hdf committed:
Changed the command-line option for disabling named pipes from -p to -N to be consistent with other tests.

kyang2014 commented on test/vfd_swmr_zoo_writer.c:
```diff
@@ -74,335 +57,437 @@ print_cache_hits(H5C_t H5_ATTR_UNUSED *cache)
+    fprintf(stderr,   "  -S: do not use VFD SWMR\n");
+    fprintf(stderr,   "  -a: run all tests, including variable-length data\n");
+    fprintf(stderr,   "  -e: print error stacks\n");
+    fprintf(stderr,   "  -l secs: maximal seconds for reader to validate the zoo\n");
```
This "-l secs: maximal seconds for reader to validate the zoo\n" seems not accurate. From the code it seems that it is "maximal number of ticks for reader to validate the zoo\n"

raylu-hdf replied:
I corrected it.

kyang2014 commented on test/testvfdswmr.sh.in:
```diff
@@ -571,24 +571,21 @@ done
-	STDOUT_PATH="./fifo" catch_out_err_and_rc vfd_swmr_zoo_reader \
-	    ../vfd_swmr_zoo_reader -q -W &
+	catch_out_err_and_rc vfd_swmr_zoo_reader \
+	    ../vfd_swmr_zoo_reader -l 4 -q &
```
Better to add a comment "the expected maximal number of ticks for reader to validate"

raylu-hdf replied:
I did it.

kyang2014 reviewed (commented):
The main() was long before. With the "named pipe" added, the main() function becomes extremely long(more than 400 lines). Is it possible to divide into a few subroutines?

gnuoyd commented:
The changes in the PR appear to substantially modify the behavior of the test without providing a clear rationale. I am concerned that with the new behavior, there are very few opportunities for the reader and writer to concurrently access the HDF5 file and the metadata file. That is, it's a weaker test than before.

A couple of questions:

What end result are the changes supposed to bring about?

Is there a smaller set of changes that produces the same end result?

raylu-hdf commented:
> The changes in the PR appear to substantially modify the behavior of the test without providing a clear rationale. I am concerned that with the new behavior, there are very few opportunities for the reader and writer to concurrently access the HDF5 file and the metadata file. That is, it's a weaker test than before.
> 
> A couple of questions:
> 
> What end result are the changes supposed to bring about?
> 
> Is there a smaller set of changes that produces the same end result?

The team discussed the instrumentation part in the zoo test and the driver.  The consensus was that it would be difficult to make it portable, especially on Windows.  Removing it will also make the test easier to maintain.  The concern for the synchronization between the writer and reader is mainly on the time difference between the writer's finishing zoo and the reader's finishing validation.

raylu-hdf committed:
Refactored the code to reduce the length of the main function.

raylu-hdf commented:
> The main() was long before. With the "named pipe" added, the main() function becomes extremely long(more than 400 lines). Is it possible to divide into a few subroutines?

I did my best to refactor it and reduced the length of the main function.

raylu-hdf commented:
I believe I have addressed all of your concerns and comments.  Is this PR good to merge?
@kyang2014 @vchoi-hdfgroup

kyang2014 commented:
The change looks good to me.

vchoi-hdfgroup commented:
Minor suggestions: 
(1) Move the closing of named pipes to a subroutine: e.g. close_open_named_pipes()
(2) [If (use_named_pipe && create_open_named_pipes() < 0).....]  may simplify the subroutine: create_open_named_pipes()

raylu-hdf committed:
Minor changes: refactored to simplify the code.

raylu-hdf committed:
Minor change: removed an unused line.

raylu-hdf commented:
> Minor suggestions:
> (1) Move the closing of named pipes to a subroutine: e.g. close_open_named_pipes()
> (2) [If (use_named_pipe && create_open_named_pipes() < 0).....] may simplify the subroutine: create_open_named_pipes()

I fixed them.
